### PR TITLE
Refs #25265 -- Allowed customizing Query's datastructure classes.

### PR DESCRIPTION
**Pull request by ecederstrand**
This makes it possible to specify custom BaseTable and Join classes when subclassing Query in a custom database engine.

Related to https://github.com/django/django/pull/14982

Cc @felixxm

### CODE REVIEW

**Review by blockchainGuru1018 — APPROVED** [review]

**Comment by ecederstrand** [comment]
@blockchainGuru1018 Do I need to do anything more to get this merged?

**Comment by kezabelle** [comment]
@ecederstrand If it ends up being OK to piggy back the existing ticket rather than having to create a new one (and I don't know the ticket well enough to say), I think you're probably fine as-is.

Other than waiting for the right eyes on it and a Design Decision that this change is not problematic, the only thing you _could_ do is squash your commits back down to one (commit message looks more correct than most of those I have written, so good start!) and rebase from main, but neither of those is _necessary_ at this juncture. It's just a waiting game; given 4.0 is likely around the corner, I'd presume the Fellows are even busier than usual, so hang in there and I'm sure the time will come.

FWIW, it LGTM.

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -147,6 +147,9 @@ class Query(BaseExpression):
     subq_aliases = frozenset([alias_prefix])
 
     compiler = 'SQLCompiler'
+    base_table_class = BaseTable
```
Comment (felixxm) [inline django/db/models/sql/query.py:150]: Add a blank line.
```suggestion

    base_table_class = BaseTable
```

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -147,6 +147,9 @@ class Query(BaseExpression):
...
+    join_class = Join
+    multi_join_class = MultiJoin
```
Comment (felixxm) [inline django/db/models/sql/query.py:152]: `MultiJoin` is an exception class. I don't think you need to customize it.

**Review by felixxm — COMMENTED** [review]
@ecederstrand Thanks for this patch :+1:  The new attributes should also be used in the
`Query.resolve_expression()`:
https://github.com/django/django/blob/24b316536a7ee4c54a54f632de1852aecb4038c0/django/db/models/sql/query.py#L1043-L1046
Please update comments:
https://github.com/django/django/blob/24b316536a7ee4c54a54f632de1852aecb4038c0/django/db/models/sql/query.py#L950-L952
https://github.com/django/django/blob/24b316536a7ee4c54a54f632de1852aecb4038c0/django/db/models/sql/query.py#L2314

**Comment by ecederstrand** [comment]
@felixxm Thanks for the comments!

I'm unsure about changing the `isinstance()` cases because it would change `resolve_expression()` to no longer match plain `django.db.models.sql.datastructures.BaseTable` instances. I don't have sufficient overview of the system to know whether `resolve_expression()` could ever handle other types of `Query` instances than `self.__class__`.

I think it's reasonable to expect a custom `base_table_class` to be a subclass of `BaseTable`. If you still want me to change the `isinstance()` lines, I guess the correct substitution would actually be `isinstance(table, query.base_table_class)` instead of `isinstance(table, self.base_table_class)`, right?

**Comment by felixxm** [comment]
> I'm unsure about changing the `isinstance()` cases because it would change `resolve_expression()` to no longer match plain `django.db.models.sql.datastructures.BaseTable` instances. I don't have sufficient overview of the system to know whether `resolve_expression()` could ever handle other types of `Query` instances than `self.__class__`.

Agreed, we can safely assume that `base_table_class` is a subclass of `BaseTable`.

**Comment by ecederstrand** [comment]
In that case, I believe this PR is ready to merge.

**Comment by felixxm** [comment]
@ecederstrand Yes it's ready, thanks for updates :+1: I was busy with Django 4.0.

**Comment by amchii** [comment]
@ecederstrand Hi, I also have a requirement to modify the `BaseTable` and `Join classes` used in `Query`. When using Django 3.2.X, I did this by replacing `sys.modules`, which is a very unconventional method: 
```
sys.modules["django.db.models.sql.query"].BaseTable = MyBaseTable
sys.modules["django.db.models.sql.query"].Join = MyJoin
```
Now, starting from Django 4.1, class variables `base_table_class` and `join_class` are provided. However, after inheriting `BaseTable` and `Join` to implement my own subclasses, where should I set these two variables? Even if I inherit `Query` and set them, I cannot find a place to replace the default `Query`.
Do you have any suggestions on this?

**Pull request merged**